packages/next-swc/crates/next-core/js/src/overlay/internal/container/Errors.tsx

```diff
@@ -130,221 +130,287 @@ const enum DisplayState {
 
-  React.useEffect(() => {
-    if (defaultTab === TabId.TurbopackErrors) {
-      setSelectedTab(TabId.TurbopackErrors);
-    }
-  }, [defaultTab]);
+  const tabs = TABS.map((tab) => ({
+    id: tab.id,
+    autoOpen: tab.autoOpen,
+    severity: tab.severity,
+    icon: tab.icon,
+    message: tab.message,
+    pluralTitle: tab.pluralTitle,
+    singularTitle: tab.singularTitle,
+    as: tab.as,
+    items: tab.items({ issues, readyErrors }),
+  })).filter((tab) => tab.items.length > 0)
 
-  const onlyHasWarnings = !hasErrors && !hasTurbopackErrors;
+  // Selected tab, null means it's closed
+  const [selectedTab, setSelectedTab] = React.useState<string | null>(null)
 
-  const [stateDisplayState, { fullscreen, minimize, hide }] = useDisplayState(
-    onlyHasWarnings ? DisplayState.Minimized : DisplayState.Fullscreen
-  );
-  let displayState = stateDisplayState;
+  // Toast is hidden
+  const [hidden, setHidden] = React.useState(false)
 
-  if (!isClosable) {
-    displayState = DisplayState.Fullscreen;
-  }
+  // Already seen issue ids, to auto open the dialog on new errors
+  const [seenIds, setSeenIds] = React.useState(() => new Set())
+
+  React.useEffect(() => {
+    const newSeenIds = new Set()
+    let change = false
+    let autoOpen = false
+    // When the selected tab disappears we will go to another important tab or close the overlay
+    if (selectedTab && !tabs.some((tab) => tab.id === selectedTab)) {
+      const otherImportantTab = tabs.find((tab) => tab.autoOpen)
+      if (otherImportantTab) {
+        setSelectedTab(otherImportantTab.id)
+      } else {
+        setSelectedTab(null)
+      }
+    } else {
+      autoOpen = true
+    }
+    // When there is a new item we open the overlay when autoOpen is set
+    for (const tab of tabs) {
+      for (const item of tab.items) {
+        newSeenIds.add(itemHash(item))
+        if (!seenIds.has(itemHash(item))) {
+          change = true
+          setHidden(false)
+          if (autoOpen && tab.autoOpen) {
+            setSelectedTab(tab.id)
+          }
+        }
+      }
+    }
+    if (change || newSeenIds.size !== seenIds.size) setSeenIds(newSeenIds)
+  }, [selectedTab, tabs, seenIds])
 
   // This component shouldn't be rendered with no errors, but if it is, let's
   // handle it gracefully by rendering nothing.
-  if (!hasErrors && !hasTurbopackWarnings && !hasTurbopackErrors) {
-    return null;
+  if (tabs.length === 0) {
+    return null
   }
 
-  if (displayState === DisplayState.Hidden) {
-    return null;
+  if (hidden) {
+    return null
   }
 
-  if (displayState === DisplayState.Minimized) {
+  if (selectedTab === null || !tabs.some((tab) => tab.id === selectedTab)) {
+    const errors = tabs.reduce(
+      (sum, tab) => sum + (tab.severity === 'error' ? tab.items.length : 0),
+      0
+    )
+    const warnings = tabs.reduce(
+      (sum, tab) => sum + (tab.severity === 'warning' ? tab.items.length : 0),
+      0
+    )
```

Comment:
the whole state of this is a pretty big mess, maybe it would be better of as a reducer 🤔

I could look into refactoring that